### PR TITLE
API: Replace special case of deprecated RuntimeIOException

### PR DESCRIPTION
**Pull request by ajantha-bhat**
9 places in the code use deprecated `RuntimeIOException` without an actual `IOException` cause (which are considered as special case).

- Introduce `FileHandlingException` which extends `RuntimeException` and replace 6 (out of 9) places with `FileHandlingException`
- Use `IOException`, `RuntimeException` directly for 2 cases where new `FileHandlingException` is unnecessary
- Remove throwing `RuntimeIOException` from one place where throwing this exception was unnecessary.

This PR partly helps in removing the deprecated `RuntimeIOException`

Derived from an end-to-end PR: https://github.com/apache/iceberg/pull/4776

### CODE REVIEW

**Comment by ajantha-bhat** [comment]
cc: @danielcweeks, @rdblue

File: api/src/main/java/org/apache/iceberg/Files.java [hunk 1]
```diff
@@ -57,7 +57,7 @@ public PositionOutputStream create() {
       }
 
       if (!file.getParentFile().isDirectory() && !file.getParentFile().mkdirs()) {
```
Comment (danielcweeks) [inline api/src/main/java/org/apache/iceberg/Files.java:59]: Maybe we could just remove these checks.  I feel like this is pretty bad error handling since we check two different conditions and don't report which or why we couldn't create the directory.

What if instead we just allowed the underlying exception to propagate up?

File: core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java [hunk 2]
```diff
@@ -85,7 +86,7 @@ public long length() {
         throw new RuntimeIOException(e, "Failed to get stream length");
       }
     }
-    throw new RuntimeIOException("Failed to get stream length: no open stream");
+    throw new FileHandlingException("Failed to get stream length: no open stream");
```
Comment (danielcweeks) [inline core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java:89]: Could we change this to a precondition check as opposed to file exception?

File: core/src/main/java/org/apache/iceberg/avro/AvroIO.java [hunk 3]
```diff
@@ -178,7 +179,7 @@ static long findStartingRowPos(Supplier<SeekableInputStream> open, long start) {
           SYNC_READER.read(decoder, blockSync);
 
           if (!Arrays.equals(fileSync, blockSync)) {
-            throw new RuntimeIOException("Invalid sync at %s", nextSyncPos);
+            throw new FileHandlingException("Invalid sync at %s", nextSyncPos);
```
Comment (danielcweeks) [inline core/src/main/java/org/apache/iceberg/avro/AvroIO.java:182]: We might be able to find a better exception for this.  I'm not clear on what this sync check is, but maybe there's something more appropriate we can use.

File: api/src/main/java/org/apache/iceberg/Files.java (hunk 1 above)
Comment (danielcweeks) [reply api/src/main/java/org/apache/iceberg/Files.java:59]: It looks like there are only four places not in test where it's being used and two or three of those are in catch blocks.  I think we could probably find a better set of exceptions to use in the few places where it is thrown and remove this entirely.

File: core/src/main/java/org/apache/iceberg/avro/AvroIO.java (hunk 3 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/avro/AvroIO.java]: Avro itself throws direct IOException for invalid sync cases as we can see here

https://github.com/apache/avro/blob/5548c9dff7a3bd851eacc791b52b21804f9f8408/lang/java/avro/src/main/java/org/apache/avro/file/DataFileStream.java#L330-L332

So, I too changed this now to use just IOException

File: core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java (hunk 2 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java]: ACK. Changed it.

File: api/src/main/java/org/apache/iceberg/Files.java (hunk 1 above)
Comment (ajantha-bhat) [reply api/src/main/java/org/apache/iceberg/Files.java]: I agree that there is no point in throwing the exception without mentioning why it has failed. 

I think we still need to create a directory for the parent path right? So, I kept that code and removed the exception part.

**Comment by ajantha-bhat** [comment]
some flaky error (seen randomly in other builds also). 
Maybe we need to enable `--stacktrace` in CI to figure it out

```
Execution failed for task ':iceberg-spark:iceberg-spark-3.3_2.12:checkstyleMain'.
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
> A failure occurred while executing org.gradle.api.plugins.quality.internal.CheckstyleAction

   > An unexpected error occurred configuring and executing Checkstyle.
See https://docs.gradle.org/8.0.1/userguide/command_line_interface.html#sec:command_line_warnings
      > java.lang.Error: Error was thrown while processing /home/runner/work/iceberg/iceberg/spark/v3.3/spark/src/main/java/org/apache/iceberg/spark/source/SparkPartitioningAwareScan.java
499 actionable tasks: 499 executed
```

re-triggering the build

**Comment by ajantha-bhat** [comment]
@danielcweeks: Thanks for the quick review. I have addressed all the comments. please check.

**Comment by nastra** [comment]
Sorry I'm quite late to the discussion around deprecating `RuntimeIOException`, but I was wondering whether we could get by without having to introduce `FileHandlingException`? 
How critical is it to have a special `RuntimeException` class to indicate that something went wrong, or could we make things work by directly using `RuntimeException`?

**Comment by ajantha-bhat** [comment]
> How critical is it to have a special RuntimeException class to indicate that something went wrong, or could we make things work by directly using RuntimeException?

I think Iceberg follows a style of wrapping the `RuntimeException` with a class to indicate that Iceberg is aware of this exception. So, I followed the same style. 

In the original PR #4776, @rdblue also suggested using some other exception than `IOException`. So, I  assumed it is `RuntimeException` and wrapped it.

**Comment by ajantha-bhat** [comment]
Hi, Is the community still interested in removing this deprecated `RuntimeIOException`? 
I think this PR didn't get enough attention last time.  
This is trivial stuff. But I think it will be helpful in maintaining codebase clean in the long run. 

cc: @danielcweeks, @nastra, @Fokko , @rdblue, @jackye1995

File: core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java [hunk 4]
```diff
@@ -78,14 +78,13 @@ public Metrics metrics() {
...
-        return stream.getPos();
-      } catch (IOException e) {
-        throw new RuntimeIOException(e, "Failed to get stream length");
-      }
+    Preconditions.checkNotNull(stream, "Failed to get stream length: no open stream");
+
+    try {
+      return stream.getPos();
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Failed to get stream length");
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java:86]: is this change actually related to this particular PR?

File: core/src/main/java/org/apache/iceberg/avro/AvroIO.java [hunk 5]
```diff
@@ -178,7 +178,7 @@ static long findStartingRowPos(Supplier<SeekableInputStream> open, long start) {
           SYNC_READER.read(decoder, blockSync);
 
           if (!Arrays.equals(fileSync, blockSync)) {
-            throw new RuntimeIOException("Invalid sync at %s", nextSyncPos);
+            throw new IOException(String.format("Invalid sync at %s", nextSyncPos));
```
Comment (nastra) [inline core/src/main/java/org/apache/iceberg/avro/AvroIO.java:181]: this change doesn't seem to be related to the scope of this PR?

File: core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java [hunk 6]
```diff
@@ -78,14 +78,13 @@ public Metrics metrics() {
 
   @Override
   public long length() {
-    if (stream != null) {
-      try {
-        return stream.getPos();
-      } catch (IOException e) {
-        throw new RuntimeIOException(e, "Failed to get stream length");
-      }
+    Preconditions.checkNotNull(stream, "Failed to get stream length: no open stream");
```
[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java:81]: Yes, I changed this to `IllegalStateException`  based on Dan's comment last time.
https://github.com/apache/iceberg/pull/6887#discussion_r1114846404

File: core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java (hunk 4 above)
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java:86]: Yes. It was also for a special case handling (RuntimeIO exception without IO exception cause).

Updated the PR title and description. So, it makes sense now.

Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java:86]: Sorry. I got confused. 

There was two `RuntimeIOException` from this length() method. Now, one is replaced with  `IllegalStateException` as per Dan's comment. So, there will still be one `RuntimeIOException` (which will be replaced with `UncheckedIOException` in the main PR as it has a cause of `IOException`)

File: core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (nastra) [reply core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java:81]: did you potentially forget to push? Because I don't see `Preconditions.checkState(...)` here

[the remark this answers is not in the document]
Comment (ajantha-bhat) [reply core/src/main/java/org/apache/iceberg/avro/AvroFileAppender.java]: Aah, I used `checkNotNull` before instead of `checkState` before as it was just replacing NULL check. 

But yeah. It is better to throw  `IllegalStateException` instead of `NPE`.  I have updated now.

**Comment by ajantha-bhat** [comment]
PR is ready @danielcweeks, @nastra, @Fokko

**Review by jbonofre — APPROVED** [review]
LGTM

File: api/src/main/java/org/apache/iceberg/exceptions/FileHandlingException.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.exceptions;
+
+import com.google.errorprone.annotations.FormatMethod;
+
+/** Exception raised when unexpected behaviour happens in file operations. */
+public class FileHandlingException extends RuntimeException {
```
Comment (Fokko) [inline api/src/main/java/org/apache/iceberg/exceptions/FileHandlingException.java:24]: I'm confused by the end goal, we're getting rid of the `RuntimeIOException`, but are creating a new exception, namely `FileHandlingException`. Shouldn't we just replace the `FileHandlingException` with `UncheckedIOException`? I think that was the goal looking at https://github.com/apache/iceberg/pull/1154

Comment (ajantha-bhat) [reply api/src/main/java/org/apache/iceberg/exceptions/FileHandlingException.java:24]: `FileHandlingException` is for `RuntimeIOException` without an underlaying `IOException` cause.
For this kind, throwing `UncheckedIOException`  when there is no `IOException` underneath is very odd. Hence throwing a `RuntimeException` (Since Iceberg wraps `RuntimeException`  to be a known exception), by introducing new `FileHandlingException`.

Ryan only suggested to throw different exception for this kind (https://github.com/apache/iceberg/pull/4776#discussion_r875038397) long back instead of  `UncheckedIOException`

https://github.com/apache/iceberg/pull/4776#discussion_r875038397

Comment (ajantha-bhat) [reply api/src/main/java/org/apache/iceberg/exceptions/FileHandlingException.java:24]: Regarding end goal. 

`RuntimeIOException` has two kinds occurrence in code.
a) One is with the cause of `IOException`.  (102 places). --> It is replaced blindly by `UncheckedIOException`  -> PR #4776

b) One without any exception cause. (9 places)  -> PR #6887
Explained above.

**Comment by ajantha-bhat** [comment]
Closing as we concluded to undo deprecation.
Fixed by https://github.com/apache/iceberg/pull/5640
